Hv Clean Architecture

Evgen1994-McD opened the pull request.

Evgen1994-McD committed:
начало рефа

Evgen1994-McD committed:
в процессе распределения по пакетам

Evgen1994-McD committed:
До конца шага 3, до начала шага 4

Evgen1994-McD committed:
Использование

Evgen1994-McD committed:
работает поиск

Evgen1994-McD committed:
апдейт

Evgen1994-McD committed:
FavoritTracks настроил передачу через ДТО

Evgen1994-McD committed:
FavoritTracks настроил передачу через ДТО

Evgen1994-McD committed:
Настраиваю Интерактор для фаворит треков

Evgen1994-McD committed:
Настроено поведение приложения при отсутсвии интернета/ отсутствии резульатта. Настроена логика с фаворит треками

Evgen1994-McD committed:
Удален FavoriteTrackAdapter.kt как и фаворит вью холдер за ненадобностью. Вся логика реализована в обычном адаптере, и зачем я его создавал изначально? хм не понимаю

Evgen1994-McD committed:
Практически реализовал архитектуру в окне настроек. Нужно немного доделать. Сделан интерктор на основные действия, Но не сделал ТЕМУ!

Evgen1994-McD committed:
update

Evgen1994-McD committed:
update

Evgen1994-McD committed:
update!!

Evgen1994-McD committed:
update!!

Evgen1994-McD committed:
update!!

Evgen1994-McD committed:
update!!

Evgen1994-McD committed:
переделать FavoriteTrackInteractorImpl.kt там не нужен консумер

Evgen1994-McD committed:
TrackInteractor - нужно убрать Дебаунс поиск внутрь него

Evgen1994-McD committed:
логика поиска полностью перенесена в интерактор включая Дебаунс поиск на 2 секунды

Evgen1994-McD committed:
Пофиксил названия, интеракторов и Юз кейсов

Evgen1994-McD committed:
IntentAndStartActivity - перенёс метод в Трек интерактор

Evgen1994-McD committed:
Создан интерактор, коммит ДО применения его на активити

Evgen1994-McD committed:
Намучился с плеером, забыл поставтиь смену статуса после подготовки плеера в интеракторе.

Evgen1994-McD committed:
Много изменений, подготовка к сдаче работы. Дебаунс на медиа вроде не нужен, перенес листенер и другие методы по своим местам

Evgen1994-McD committed:
Применил автоформат кода ко всем файлам

Evgen1994-McD committed:
Сдача на проверку

Evgen1994-McD committed:
Сдача на проверку

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/data/repositories/SettingsReposytoryImpl.kt:
```diff
@@ -0,0 +1,77 @@
+import com.example.playlistmaker.domain.api.SettingsRepository
+
+class SettingsReposytoryImpl() : SettingsRepository {
+    override fun shareApp(context: Context): Intent {
```
такие зависимости как `Context` в репозитории лучше получать через конструктор

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/Creator.kt:
```diff
@@ -0,0 +1,81 @@
+        return FavoriteTrackRepositoryImpl(context)
+    }
+
+    fun provideFavoriteInteractor(context: Context): FavoriteTrackInteractor {
```
чтобы не передавать контекст через все слои абстракции - можно в Creator объявить поле для сохранения контекста, и получить его в `App` на `onCreate`

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/data/repositories/SettingsReposytoryImpl.kt:
```diff
@@ -0,0 +1,77 @@
+        }
+
+        return sendIntent
+//        context.startActivity(
```
Закоментированный код лучше сразу удалять. Хранить закоментированный код в проекте - плохая практика, в случае чего этот код можно восстановить с помощью истории гита

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/App.kt:
```diff
@@ -15,30 +14,33 @@ class App : Application() { // класс АПП для смены темы
 
     fun hasBooleanValue(context: Context, key: String): Boolean {
-        val sharedPref = context.getSharedPreferences(Constants.SHARED_PREF_THEME_NAME, Context.MODE_PRIVATE)
+        val sharedPref = context.getSharedPreferences(
```
Все что связано префсами и темой приложения - должно находиться в репозитории который работает с темой, 

на старте приложения должен быть только вызов интерактора который работает с темой и применяет сохраненную тему

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/domain/api/FavoriteTrackRepository.kt:
```diff
@@ -0,0 +1,23 @@
+import com.example.playlistmaker.domain.models.Track
+
+interface FavoriteTrackRepository {
+    fun addTrack(track: Track) {}
```
на самом деле будет достаточно трех публичных методов:
- сохранить список
- получить список
- очистить список

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/domain/impl/SwitchThemeUseCaseImpl.kt:
```diff
@@ -0,0 +1,52 @@
+            }
+
+            // Перезапустить активность для применения новой темы
+            recreate(activity)
```
этот вызов не нужен, более того - так мы не будем передавать контекст активити сюда

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/presentation/activity/MediaActivity.kt:
```diff
@@ -15,94 +13,49 @@ import androidx.appcompat.app.AppCompatActivity
     private var isPlaying = false // переменная статуса плеера
-
-
+private lateinit var favoriteTrackInteractorImpl : FavoriteTrackInteractor
```
можно через lazy:
`private val favoriteTrackInteractorImpl: FavoriteTrackInteractor by { Creator.provideFavoriteInteractor() }`

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/presentation/activity/SettingsActivity.kt:
```diff
@@ -0,0 +1,96 @@
+import com.google.android.material.textview.MaterialTextView
+
+class SettingsActivity : AppCompatActivity() {
+private lateinit var shareAppUseCase : ShareAppUseCase
```
можно так же через lazy - тогда объект создастся только по мере необходимости (при его непосредственном вызове)

Evgen1994-McD committed:
Поле для контекста в Creator
(исправление замечаний)

Evgen1994-McD committed:
Разобраться с этими контекстами в смене темы и других классах

Evgen1994-McD committed:
1. Сделал by lazy
2. разобрался с темой
3. В АПП только вызов интерактора, все методы перенес в репозитойри

Evgen1994-McD committed:
1. Удалил часть лишних метододов в классе FavoriteTrackRepositoryImpl.kt

ToDO : Рассмотреть вариант переноса листенера на активити, чтобы убрать последний лишний метод из репозитория

Evgen1994-McD committed:
Исправлены все недочёты на ревью, сдача проекта
